### DISCUSSION TITLE
iCub 2 Shoulder Constraints  degrees <--> radians?

### DISCUSSION THREAD

**Discussion by Woolfrey** [robotology/community#602] [Q&A] [answered]
The [iCub wiki](https://icub-tech-iit.github.io/documentation/icub_kinematics/icub-shoulders-range-of-motion/icub-shoulders-range-of-motion/#correct-limits) specifies the shoulder constraints as **Aq + b > 0** where **q** is in degrees, if I'm not mistaken.

I'm working in radians for all the other calculations. How should I convert the above constraint to radians? I tried multiplying the **A** matrix by 180/pi, but when I print the result of **Aq+b** the result seems very large (i.e. very far from the constraints). I'm not sure if this is correct or not!

Thanks in advance!

**Comment by pattacini** [comment, marked as answer]
Hi @Woolfrey 

Judging from https://github.com/robotology/icub-main/blob/master/src/libraries/iKin/src/iKinIpOpt.cpp#L155-L174, it seems that only the vector $b$ gets converted in radians, having the shoulders' joints already expressed in radians.

Therefore, the documentation is not really clear in this aspect. I'll amend it.

**Reply by pattacini** [reply]
>where q is in degrees

Actually, they're in radians.
I'll make soon a PR to support the discussion.

**Reply by Woolfrey** [reply]
EDIT: Sorry, I misread: I only need to convert the vector **b** to radians? (That would have been my second guess).

Just double-checking so I don't break the robot!

**Reply by pattacini** [reply]
>EDIT: Sorry, I misread: I only need to convert the vector b to radians? (That would have been my second guess).

Yep 👍🏻

**Reply by pattacini** [reply]
Opened up this PR:
- https://github.com/icub-tech-iit/documentation/pull/233

**Reply by pattacini** [reply]
I'll merge the PR straight away as the documentation is flawed now.

**Reply by Woolfrey** [reply]
Great, thanks! I will check it out in Gazebo tomorrow morning before moving to the real robot.

**Comment by Woolfrey** [comment]
My Cartesian controller and QP solver appear to be working fine with the added constraint. I'm getting different behaviour for the left and right arms when moving up and down:
![image](https://user-images.githubusercontent.com/62581255/203594031-e804b17d-9c4b-4067-861c-f1c7f7fce5d8.png)
I'm assuming this is because the left and right arms start in *slightly* different configurations which causes the QP solver to find different joint motion. I'm also minimizing instantaneous kinetic energy, so this might explain some of it (1 arm moving fast + 1 moving slow < 2 arms moving fast). I checked that **A*q + b** which is all non-zero as well (5 constraints for each arm, so output is 10x1 vector):
![image](https://user-images.githubusercontent.com/62581255/203595232-6071d46e-3b77-466a-9669-f0c05458244e.png)
Is this kind of shoulder motion to be expected, in your experience?

**Reply by pattacini** [reply]
The different behaviors you spotted are perfectly normal (I saw them too) as long as the primary task is deemed attained correctly.

The solver can find different configurations of the shoulders due to the different starting points and the redundancy of the manipulators unless some regularization term is explicitly enforced, for example on the height of the elbow, which in turn will constrain the search space of the solver.
